AI 챗봇 앱 [STEP 3] 벨로, 밤삭

bamsak opened the pull request:
리뷰어: @forestjae

### 팀원
| 프로필 사진 | <a href="https://github.com/bamsak"><img src="https://avatars.githubusercontent.com/u/114239407?v=4" width="90" height="90"></a> | <a href="https://github.com/JinUng41"><img src="https://avatars.githubusercontent.com/u/91656206?v=4" width=90></a> | 
| ---- | ---------- | --------- | 
| in Github | [@bamsak](https://github.com/bamsak) | [@JinUng41](https://github.com/JinUng41) 
| in SeSAC | 밤삭 | 벨로

안녕하세요, 숲재!!

잘 부탁드립니다!! 🙇‍♂️🙇‍♂️

[STEP 3]에서 기존에 구현했던 객체들의 기능을 활용하여 View와 연동하고 앱이 동작하도록 구현했습니다.

### 핵심 요구사항

- [x]  Safe Area을 고려한 오토레이아웃 구현
- [x]  Collection View의 활용
- [x]  Modern Collection View 활용
- [x]  Core Graphics와 Core Animation의 활용

## 구현 내용

### 폴더 구조

```
.
├── Application
│   ├── AppDelegate.swift
│   └── SceneDelegate.swift
├── Controllers
│   └── GPTChatRoomViewController.swift
├── Extensions
│   ├── Bundle+Extension.swift
│   ├── GPTChatRoomViewController+Extension.swift
│   ├── GPTMessageDTO+Extension.swift
│   ├── UITextView+Extension.swift
│   └── URLSession+Extension.swift
├── Models
│   ├── DTOs
│   │   ├── Implementation
│   │   │   ├── Common
│   │   │   │   └── GPTMessageDTO.swift
│   │   │   ├── GPTRequestDTO.swift
│   │   │   └── GPTResponseDTO.swift
│   │   └── Interface
│   │       └── RequestEncodable+RequestDecodable.swift
│   ├── Messages
│   │   ├── Implementation
│   │   │   ├── AssistantMessage.swift
│   │   │   ├── SystemMessage.swift
│   │   │   └── UserMessage.swift
│   │   └── Interface
│   │       └── GPTMessageable.swift
│   └── Network
│       ├── APIRequest
│       │   ├── Implementation
│       │   │   └── GPTRequest.swift
│       │   └── Interface
│       │       └── APIRequestable.swift
│       ├── Error
│       │   └── NetworkError.swift
│       └── HTTPMethod.swift
├── Protocols
│   └── URLSessionProtocol.swift
├── Resources
│   ├── Assets.xcassets
│   │   ├── AccentColor.colorset
│   │   │   └── Contents.json
│   │   ├── AppIcon.appiconset
│   │   │   └── Contents.json
│   │   └── Contents.json
│   ├── Base.lproj
│   │   └── LaunchScreen.storyboard
│   └── Info.plist
├── ServiceProvider
│   ├── Implementation
│   │   └── GPTServiceProvider.swift
│   └── Interface
│       └── ServiceProvidable.swift
├── Utilities
│   ├── JSON
│   │   ├── Implementation
│   │   │   └── JSONCoder.swift
│   │   └── interface
│   │       └── JSONCodable.swift
│   ├── Network
│   │   └── NetworkManager.swift
│   └── URLRequestConverter
│       ├── Implementation
│       │   └── URLRequestConverter.swift
│       └── Interface
│           └── URLRequestConvertible.swift
├── ViewModels
│   └── GPTViewModel.swift
└── Views
    ├── ChatBubbleView.swift
    ├── MessageCell.swift
    ├── MessageContentConfiguration.swift
    └── MessageContentView.swift
```

### 동작 화면
|채팅 입력 | Keyboard Layout Guide |
|-------------------------------------------|------------------------------------------------|
| ![Simulator Screen Recording - iPhone 15 Pro - 2024-01-16 at 19 34 59](https://github.com/tasty-code/ios-chat-bot/assets/114239407/18217a17-1f9a-4dd1-9525-1240e9d20989) | ![Simulator Screen Recording - iPhone 15 Pro - 2024-01-16 at 19 37 53](https://github.com/tasty-code/ios-chat-bot/assets/114239407/e7e80919-a2c5-4b37-b168-1eab587c287c) |

## MVVM

- `didSet`과  `closure`를 이용하여 Data Binding을 구현하였습니다.
- View에서 View Model의 클로저의 값을 정의해주며 View Model에서의 Data를 통해, Diffable DataSource의 snapshot을 구성할 수 있도록 하였습니다.

```swift
// View Model
private var messages: [GPTMessageDTO] = [] {
        didSet {
            didMessagesAppend?(messages)
        }
    }
    
var didMessagesAppend: (([GPTMessageDTO]) -> Void)?

// ViewController

init(viewModel: GPTViewModel) {
     self.viewModel = viewModel
     super.init(nibName: nil, bundle: nil)
        
     viewModel.didMessagesAppend = { [weak self] messages in
        guard let self else { return }
        configureSnapshot(with: messages)
        updateUI()
    }
}
```

## Modern Cell Configuration

```swift
private let cellRegistration: UICollectionView.CellRegistration<MessageCell, GPTMessageDTO> = {
    return UICollectionView.CellRegistration<MessageCell, GPTMessageDTO> {
        cell, indexPath, item in
        var configuration = MessageContentConfiguration()
        configuration.message = item
        cell.contentConfiguration = configuration
    }
}()
```

- CellRegistration에서 Configuration을 통해 필요한 데이터를 넘겨주는 방식을 구현하였습니다.
- 커스텀한 MessageCell, MessageContentConfiguration, MessageContentView를 구현하였습니다.
- 기존의 UICollectionViewDataSource 프로토콜을 채택하여 `cellForRowAt`에서 셀을 구성하는 방식은 셀에게 직접 필요한 데이터를 전달해 주었다면
- Cell Configuration은 ContentConfiguration을 이용하여 필요한 값을 할당하고, ContentView에서 Configuration을 통해 필요한 데이터를 뷰에 띄웁니다.
    - 이 때, ContentConfiguration은 값 타입이기 때문에 생성 비용이 저렴하며
    - UIKit이 DiffableDataSource를 통해 변경을 파악하고 필요에 따라 뷰를 효과적으로 업데이트한다고 이해하였습니다.
- 이에 따라 해당 Cell의 ContentView에서의 애니메이션 처리와 말풍선 그리는데 필요한 draw메소드의 분기처리를  통해 화면에 원하는 동작이 나올 수 있도록 했습니다.

## 궁금한 점

### 의존성 관리

현재 View에 ViewModel을 SceneDelegate에서 의존성 주입을 통해 View Mdoel 객체를 받고 있습니다.

그러다보니 SceneDelegate단에서 인스턴스가 많이 생성되고 있는데, 이러한 부분은 어떤 방식으로 관리가 되어야 하는지 궁금합니다.

의존성 주입을 관리하는 하나의 싱글톤 객체를 두어 관리하는 방식 등을 생각해 보았는데요..

이와 관련해서 숲재의 의견을 듣고 싶습니다.

```swift
// SceneDelegate

private let jsonCoder = JSONCoder()
private let networkManager = NetworkManager()
private lazy var gptServiceProvider = GPTServiceProvider(networkManager: networkManager, jsonCoder: jsonCoder)
private lazy var gptViewModel = GPTViewModel(serviceProvider: gptServiceProvider)

func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
    guard let windowScene = (scene as? UIWindowScene) else { return }
    window = UIWindow(windowScene: windowScene)
    window?.rootViewController = GPTChatRoomViewController(viewModel: gptViewModel) 
    window?.makeKeyAndVisible()
}
```

### MARK 주석 사용법

GPTChatRoomViewController와 MessageContentView에서 MARK 주석을 사용하고 있습니다.

주석을 이용해 나름 코드의 영역? 을 보다 분리한다고 생각했지만, 정확하게 어떻게 써야할 지 감이 잡히지 않아서 조언을 듣고 싶습니다!

bamsak committed:
Merge pull request #2 from bamsak/step2

Step2 merge

JinUng41 committed:
[Refactor] excute 메소드 동작 방법 수정

- GPTRequestDTO는 RequestDTOEncodable을, GPTResponseDTO는 ResponseDTODecodable을 채택하도록 구현
- 기존 `associatedType`이 아닌 RequestDTOEncodable & ResponseDTODecodable 타입을 사용하는 제네릭 함수로 구현

JinUng41 committed:
[Feat] 채팅방 UI 구성

- ViewController 이름을 GPTChatRoomViewController로 변경

JinUng41 committed:
[Chore] GPTChatRoomViewController+Extension 추가

- 기존에 뷰컨트롤러에 같이 있던 익스텐션을 파일로 분리
- 뷰컨트롤러의 익스텐션에는 프로토콜 채택 내용이 닮김.

JinUng41 committed:
[Feat] Cell Content Configuration으로 구현

JinUng41 committed:
[Refactor] UI 구성을 위한 커스텀 Cell Configuration을 수정

- configuration을 통해 필요한 데이터를 전달받도록 함.

JinUng41 committed:
[Refactor] DiffableDataSource를 위한 DTO 타입 수정

- DiffableDataSource에 사용할 객체로 GPTMessageDTO 결정
- 그에 따른 Hashable 채택 및 GPTRequestDTO 수정

JinUng41 committed:
[Feat] 주요 객체들을 연결하고 채팅 기능을 구현

- SceneDelegate에서 필요한 객체들에 대해 의존성을 주입
- ViewModel을 정의 (클로저 방식으로 데이터 바인딩)
- DiffableDataSource와 CellRegisteration을 구현
- DiffableDataSource를 위한 Section 열거형 정의

bamsak committed:
[Refactor] ViewModel 로직 수정

- label 먼저 생성 후, 들어온 데이터를 매핑하도록 로직수정

JinUng41 committed:
[Refactor] GPTMessageDTO의 role에 해당하는 타입 변경

- String에서 따로 정의한 열거형인 MessageRole 타입으로 변경

JinUng41 committed:
[Feat] 말풍선 로딩 애니메이션 구현

- API로부터 응답이 오는 동안 보여줄 로딩 애니메이션을 구현

JinUng41 committed:
[Refactor] 가독성을 위한 코드 분리

- code depth를 줄이기 위한 코드 분리

bamsak committed:
[Refactor] collectionView 스크롤 시 키보드 dismiss되도록 변경

JinUng41 committed:
Merge branch 'step3' of https://github.com/bamsak/ios-chat-bot into step3

JinUng41 committed:
[Chore] 빈 줄 삭제

bamsak committed:
[Chore] enum Section에 채택한 Hashable 제거

forestjae commented on ChatBot/ChatBot/Application/SceneDelegate.swift:
```diff
@@ -9,11 +9,16 @@ import UIKit
-
+    
+    private let jsonCoder = JSONCoder()
+    private let networkManager = NetworkManager()
+    private lazy var gptServiceProvider = GPTServiceProvider(networkManager: networkManager, jsonCoder: jsonCoder)
+    private lazy var gptViewModel = GPTViewModel(serviceProvider: gptServiceProvider)
+    
     func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
         guard let windowScene = (scene as? UIWindowScene) else { return }
         window = UIWindow(windowScene: windowScene)
-        window?.rootViewController = ViewController()
+        window?.rootViewController = GPTChatRoomViewController(viewModel: gptViewModel) 
         window?.makeKeyAndVisible()
     }
```
SceneDelegate에서 객체 초기화 코드들이 직접 있는게 싫다면, 다른 객체가 그 역할을 담당하면 됩니다.
유명한 Coordinator 패턴이라는게 있고요, 아니면 ViewController를 make 해주는 factory 같은것을 활용하면 이런 상황을 방지할 수 있어요. 네트워크 매니저나 JSONCoder같은 일반적인 객체는 싱글턴 사용을 고려해 볼 수도 있겠네요.

PR 본문에 질문 작성해주셨는데, 여기다가 답변 드려요.

bamsak replied:
감사합니다!🙇‍♂️
해당 패턴들에 대해 공부해보겠습니다.

forestjae commented on ChatBot/ChatBot/Controllers/GPTChatRoomViewController.swift:
```diff
@@ -0,0 +1,200 @@
+        
+        viewModel.didMessagesAppend = { [weak self] messages in
+            guard let self else { return }
+            configureSnapshot(with: messages)
+            updateUI()
```
여기서 self 키워드를 명시해주지 않았기 때문에, weak의 효과가 적용되지 않습니다.

추가로, guard let self else { return } 구문의 단점도 공부해보세요.

bamsak replied:
guard let self else { return } 구문의 단점에 대해 알아보았는데
클로저 자체는 RC를 올리지 않지만 옵셔널 바인딩 된 self가 RC를 증가시켜
단순 [weak self]만을 사용했을 때의 동작과는 다른 모습을 보이게 됩니다.
해당 self의 참조가 클로저 내부 동작이 다 끝날 때까지 참조를 유지하여
해당 객체가 외부에서 참조가 해제되어도 클로저 내부에서의 self 때문에 해제가 되지 않는 것을
알게 되었습니다.
이로 인해, 기존의 강한 참조에서의 순환 참조는 방지할 수 있지만,
한편으로는 클로저 내부에서 self의 호출 동작이 다 끝나고 무거운 작업을 실행 하는 경우에도,
해당 객체는 클로저 구문이 다 끝날 때 까지 메모리에 살아있게 되어
무거운 작업이 진행되는 동안 메모리 릭이 발생한다고 생각이 되는데 
저희가 이해한 방식이 맞는지 의견 부탁드립니다!

forestjae replied:
네 맞습니다. 정확히는 강한 순환 참조가 일어나지 않기 때문에, 메모리 릭이라기 보다는 지연해제라고 표현하는 점 참고해주세요~!

forestjae commented on ChatBot/ChatBot/Controllers/GPTChatRoomViewController.swift:
```diff
@@ -0,0 +1,200 @@
+    private func setConstraintsCollectionView() {
+        let safeArea = view.safeAreaLayoutGuide
+        
+        view.addSubview(chatCollectionView)
+        NSLayoutConstraint.activate([
+            chatCollectionView.topAnchor.constraint(equalTo: safeArea.topAnchor),
+            chatCollectionView.leadingAnchor.constraint(equalTo: safeArea.leadingAnchor, constant: 10),
+            chatCollectionView.trailingAnchor.constraint(equalTo: safeArea.trailingAnchor, constant: -10),
+        ])
```
메서드를 나누시려는 모습은 좋습니다.

다만, addSubView와 같이 ViewHierarchy에 영향을 주는 메서드가 곳곳에 분리되어 있으면 추후에 메서드 순서가 바뀌거나 할 때 제약을 걸때 크래시가 생길 수도 있으니 순서에 유념해야 된다는 점 참고해주세요. 그래서 일반적으로 동일한 위계의 addSubView(지금의 경우에는 view.addSubView()) 메서드들은 다 모아두는 편입니다.

JinUng41 replied:
초기에는 `setupUI()`내에서 모든 내용을 정의하였으나, 보다 명확하게 하기 위해서 분리를 시도한 결과였습니다.
하지만 말씀해주신대로 뷰 계층 구조가 복잡해 지면, 휴먼 에러를 좀 더 방지하는 차원에서 `addSubview()`를 한 곳에 묶고 호출하는 것이 안전하겠다는 생각이 들게 되었습니다.
말씀 감사합니다!!

수정 또한 `setupUI()`에서 먼저 `addSubview()`를 호출한 다음 오토레이아웃 코드가 호출되도록 수정하였습니다.

forestjae commented on ChatBot/ChatBot/Controllers/GPTChatRoomViewController.swift:
```diff
@@ -0,0 +1,200 @@
+            userInteractionStackView.topAnchor.constraint(equalTo: chatCollectionView.bottomAnchor, constant: 10),
+            userInteractionStackView.leadingAnchor.constraint(equalTo: chatCollectionView.leadingAnchor),
+            userInteractionStackView.trailingAnchor.constraint(equalTo: chatCollectionView.trailingAnchor),
+            view.keyboardLayoutGuide.topAnchor
+                .constraint(equalToSystemSpacingBelow: userInteractionStackView.bottomAnchor, multiplier: 1.0)
```
이건 어떤 의미를 가진 코드인가요? 처음봐서 여쭤봅니다.

bamsak replied:
현재 저희 코드 상으로 userInteractionStackView의 bottom Anchor를 keyboardLayoutGuide의 top Anchor에 맞추도록 설정이 되어있습니다.
위의 경우 키보드가 나타나지 않을 땐, 자동으로 safeArea의 bottom Anchor에 잡히는 것으로 알고 있습니다.
또한 키보드가 나타났을 경우, 키보드가 stackView를 가리지 않고 동적인 움직임을 가져갈 수 있도록 해주는 코드로 알고 있습니다.

forestjae commented on ChatBot/ChatBot/Controllers/GPTChatRoomViewController.swift:
```diff
@@ -0,0 +1,200 @@
+        return dataSource
+    }
+    
+    private func configureSnapshot(with messages: [GPTMessageDTO]) {
+        var snapshot = dataSource.snapshot()
+        snapshot.deleteAllItems()
+        snapshot.appendSections([.main])
+        snapshot.appendItems(messages)
+        dataSource.apply(snapshot, animatingDifferences: true)
+    }
```
매번 전체 아이템을 지우고, 섹션을 추가하는 작업을 하고 있는데 현재 스냅샷을 꺼내오는 이유가 없지 않을까요?

bamsak replied:
snapshot을 꺼내온 후 delete해줄 필요 없이 snapshot을 생성해주는 방식으로 변경했습니다.

forestjae commented on ChatBot/ChatBot/Controllers/GPTChatRoomViewController.swift:
```diff
@@ -0,0 +1,200 @@
+        viewModel.fetch(userInput: userMessageText)
+        
+        userInputTextView.resignFirstResponder()
+        userInputTextView.text = nil
+        userInputTextView.insertText("")
```
text를 nil로 만들고 공백문자를 넣어주는 이유가 궁금해요

JinUng41 replied:
`insertText("")`를 통해 뷰컨트롤러가 채택한 UITextViewDelegate의 메서드인 `textViewDidChange`가 호출되어 초기 크기로 돌아갈 수 있도록 하기 위함이었습니다.

forestjae commented on ChatBot/ChatBot/Extensions/GPTChatRoomViewController+Extension.swift:
```diff
@@ -0,0 +1,33 @@
+
+extension GPTChatRoomViewController: UITextViewDelegate {
+    func textViewDidChange(_ textView: UITextView) {
+        guard textView.contentSize.height < view.frame.height * 0.15
+        else {
+            textView.isScrollEnabled = true
+            return
+        }
```
이러한 예외처리가 들어간 배경이 궁금하네요

JinUng41 replied:
내용이 입력될 때 임의의 최대 크기보다는 커지지 않으면서, 스크롤이 가능한 텍스트 뷰를 구현하기 위한 조건으로 생각하고 구현한 코드입니다.

forestjae commented on ChatBot/ChatBot/Extensions/GPTChatRoomViewController+Extension.swift:
```diff
@@ -0,0 +1,33 @@
+    }
+}
+
+extension GPTChatRoomViewController {
+    enum Section {
+        case main
+    }
+}
```
같은 extension에 있기에는 성격이 많이 다른 것 같습니다.

bamsak replied:
extension으로 빼지 않고 ViewController 내부의 nested type으로 변경하였습니다.

forestjae commented on ChatBot/ChatBot/Extensions/GPTMessageDTO+Extension.swift:
```diff
@@ -0,0 +1,18 @@
+        return lhs.uuid == rhs.uuid
+    }
+    
+    func hash(into hasher: inout Hasher) {
```
hash 메서드는 어떤 역할을 하나요?

bamsak replied:
hashable하게 관리할 value를 해당 메서드를 통해, 지정한 값으로 hash 주소를 만들어 hash table에 지정한 값의 주소로 value를 저장하는 역할을 합니다!

forestjae commented on ChatBot/ChatBot/ViewModels/GPTViewModel.swift:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+    
+    var didMessagesAppend: (([GPTMessageDTO]) -> Void)?
```
바인딩을 위해 클로저를 활용해 주셨는데, 다른 바인딩 방법에는 어떤게 있을까요?

JinUng41 replied:
클로저 이외에 Observable 클래스를 직접 구현하는 것과 Combine을 이용한 데이터 바인딩 방법이 있는 것으로 알고 있습니다.
클로저를 사용한 이유는 저희가 MVVM을 적용해본 첫 프로젝트여서, 가장 간단한 구현 방식인 클로저를 이용하여 구현해 보았습니다.

forestjae commented on ChatBot/ChatBot/ViewModels/GPTViewModel.swift:
```diff
@@ -0,0 +1,51 @@
+    }
+    
+    private func addResponseMessage(_ messageDTO: GPTMessageDTO) {
+        Task { @MainActor in
```
여기서 @MainActor 는 어떤 역할을 하나요?

bamsak replied:
Task구문이 thread를 보장하지 않기 때문에,
@MainActor의 경우, DispatchQueue.main처럼 main thread에서 동작함을 보장해줍니다.
저희 코드는 messages 배열의 값이 변경으로 인해 UI의 업데이트가 이루어집니다.
UI 업데이트 관련 로직은 main thread에서 동작해야 하므로 해당 Task구문에서 @MainActor를 사용했습니다.

forestjae commented on ChatBot/ChatBot/Views/ChatBubbleView.swift:
```diff
@@ -0,0 +1,128 @@
+    
+    var direction: Direction {
+        didSet {
+            setNeedsDisplay()
```
이 메서드가 호출되면 어떤 일이 일어나나요?

bamsak replied:
View에 변경사항이 있다면, 시스템에 View를 다시 그려야 한다는 것을 알려, 
다음 drawing cycle에 View의 draw메소드가 호출되도록 합니다.
따라서 해당 메소드를 직접 호출하여 View를 다시 그리도록 하는 동작을 할 수 있습니다.

forestjae commented on ChatBot/ChatBot/Views/ChatBubbleView.swift:
```diff
@@ -0,0 +1,128 @@
+    }
+    
+    override func draw(_ rect: CGRect) {
+        switch direction {
+        case .right:
+            drawBubbleToRight()
+        case .left:
+            drawBubbleToLeft()
+        }
```
👍

forestjae commented on ChatBot/ChatBot/Views/MessageContentView.swift:
```diff
@@ -0,0 +1,153 @@
+            chatBubble.layer.addSublayer(dotLayer)
+            
+            let animation = CAKeyframeAnimation(keyPath: "transform.scale")
+            animation.values = [1.0, 1.2, 1.0]
+            animation.keyTimes = [0, 0.5, 1]
+            animation.duration = animationDuration
+            animation.beginTime = CACurrentMediaTime() + (animationDuration / Double(dotsCount) * Double(index))
+            animation.repeatCount = .infinity
+            animation.timingFunction = CAMediaTimingFunction(name: .easeInEaseOut)
+            dotLayer.add(animation, forKey: "loading")
```
👍

forestjae commented:
고생 많으셨습니다. 상당히 복잡한 뷰를 그리는데 잘 해 주셨어요. 👍 

문의하신 내용은 코멘트로 남겼습니다.

bamsak committed:
[Refactor] 마이너한 리뷰 반영

- addSubView 호출 위치 변경
- snap을 생성하도록 변경
- Section enum 코드 위치 수정
- animation을 위한 autolayout priority 변경

bamsak committed:
[Refactor] API통신을 위한 기존의 모델 변경

- struct가 아닌 enum으로 URLRequest만들도록 구현
- URLRequestCoverter, JSONCoder객체 삭제

bamsak committed:
[Refactor] 바뀐 API모델에 따른 코드 수정

- ServiceProvider를 GPT에 종속적이지 않게 구현
- encoding, decoding 기능을 위한 각 encoder, decoder에 extension으로 구현

JinUng41 committed:
[Chore] 코드 컨벤션에 맞게 수정

- 빈 줄 삭제 등

JinUng41 commented on ChatBot/ChatBot/Models/API/APIEndPoint/Interface/APIEndPoint.swift:
```diff
@@ -0,0 +1,45 @@
+
+import Foundation
+
+protocol APIEndPoint {
+    var scheme: String { get }
+    var host: String { get }
+    var path: String { get }
+    var httpMethod: HTTPMethod { get }
+    var headerFields: [String: String]? { get throws }
+    var encodableBody: Encodable? { get }
+    
+    func makeURLRequest(with encoder: DataEncodable) throws -> URLRequest
+}
+
+extension APIEndPoint {
+    func makeURLRequest(with encoder: DataEncodable) throws -> URLRequest {
+        let url = try self.makeBaseURL()
+        var urlRequest = URLRequest(url: url)
+        urlRequest.httpMethod = "\(self.httpMethod)"
+        urlRequest.allHTTPHeaderFields = try self.headerFields
+        if let encodableBody = self.encodableBody {
+            urlRequest.httpBody = try encoder.encode(from: encodableBody)
+        }
+        return urlRequest
+    }
+    
+    private func makeBaseURL() throws -> URL {
+        var urlComponents = URLComponents()
+        urlComponents.scheme = self.scheme
+        urlComponents.host = self.host
+        urlComponents.path = self.path
+        guard let url = urlComponents.url
+        else {
+            throw APIError.badURL
+        }
+        
+        return url
+    }
+}
```
기존의 프로토콜 + 구조체 방식을 프로토콜 + 열거형 방식으로 변경하였습니다.
모든 API 통신을 위한 프로토콜을 정의하였는데요.
열거형의 특정 API가 해당 프로토콜을 채택하여 필요한 구성요소에 대한 내용들을 정의할 수 있도록 하였습니다.
extension을 통해 스스로 URL과 URLRequest를 만들어 낼 수 있도록 하였습니다.

forestjae replied:
잘 구현해주셨네요 👍

JinUng41 commented on ChatBot/ChatBot/Models/API/APIEndPoint/Implementation/GPTEndPoint.swift:
```diff
@@ -0,0 +1,60 @@
+
+import Foundation
+
+enum GPTEndPoint {
+    case chatbot(body: RequestDTOEncodable)
+}
+
+extension GPTEndPoint: APIEndPoint {
+    var scheme: String {
+        switch self {
+        case .chatbot:
+            return "https"
+        }
+    }
```
열거형인 GPTEndPoint에서 APIEndPoint를 채택하여, URLRequest를 만들기 위해 필요한 구성요소들을 API 케이스에 따라 다르게 가질 수 있도록 하였습니다.
현재는 챗봇이라는 기능만이 존재하기 때문에 case가 1개지만 추후 다른 GPT API 통신이 필요하게 될 때, 이러한 열거형의 방식이 관리하기에 보다 수월할 것이라고 저희는 생각하고 있습니다.
이와 관련하여 숲재의 의견을 들려주시면 감사하겠습니다.~👍👍

forestjae replied:
매우 좋은 방법인 것 같습니다 👍

JinUng41 commented on ChatBot/ChatBot/Models/API/APIEndPoint/Implementation/GPTEndPoint.swift:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+    
+    var headerFields: [String : String]? {
+        get throws {
+            switch self {
+            case .chatbot:
+                guard let apiKey = Bundle.main.gptAPIKey else { throw APIError.notFoundAPIKey }
+                return ["Authorization": "Bearer \(apiKey)",
+                        "Content-Type" : "application/json"]
+            }
+        }
+    }
```
챗봇 API 통신에 필요한 API KEY에 대한 오류 처리를 다음과 같이 구현해 보았습니다.
URLRequest에 필요한 헤더 필드에 대한 내용을 가져올 때, 열거형을 생성하는 외부에서 API KEY를 넣어주기 보다는 열거형 내부에서 오류를 throws 하는 방법이 더 나아 보인다고 판단하였습니다.

forestjae replied:
생각해보지 못했는데, 좋은 방법이네요.

bamsak commented on ChatBot/ChatBot/Utilities/ServiceProvider/Implementation/ServiceProvider.swift:
```diff
@@ -0,0 +1,30 @@
+
+import Foundation
+
+final class ServiceProvider: ServiceProvidable {
+    private let networkManager: NetworkManager
+    private let encoder: DataEncodable
+    private let decoder: DataDecodable
+    
+    init(networkManager: NetworkManager,
+         encoder: DataEncodable = JSONEncoder(), 
+         decoder: DataDecodable = JSONDecoder()
+    ) {
+        self.networkManager = networkManager
+        self.encoder = encoder
+        self.decoder = decoder
+    }
+    
+    func excute<E: APIEndPoint, D: ResponseDTODecodable>(for requestDTO: E) async throws -> D {
+        let urlRequest = try requestDTO.makeURLRequest(with: encoder)
+        let data = try await networkManager.fetchData(with: urlRequest)
+        let responseDTO = try decoder.decode(to: D.self, from: data)
+        return responseDTO
+    }
+}
```
encode와 decode를 프로토콜로 추상화 하여, 현재 필요한 JSONEncoder와 JSONDecoder 해당 프로토콜을 채택하여 의존성을 해당 프로토콜로 가지는 방식으로 변경 하고 기존의 JSONCoder객체를 삭제하는 방식을 적용시켰습니다.
앞에서 APIEndPoint를 구현함으로서 ServiceProvider가 GPT에 종속적이지 않고 범용적으로 사용할 수 있도록 변경했습니다.

forestjae replied:
좋습니다 👏 👏 👏 👏
